metadata_writer_nne.c

```diff
@@ -94,132 +309,725 @@ static uint8_t md_nne_handle_gps_event(struct md_writer_nne *mwn,
+{
+    struct nne_value value;
+    value.type = type;
+    switch(type)
+    {
+        case NNE_TYPE_NULL:
+            break;
+        case NNE_TYPE_INT8:
+            value.u.v_int8 = *(int8_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_UINT8:
+            value.u.v_uint8 = *(uint8_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_INT16:
+            value.u.v_int16 = *(int16_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_UINT16:
+            value.u.v_uint16 = *(uint16_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_INT32:
+            value.u.v_int32 = *(int32_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_UINT32:
+            value.u.v_uint32 = *(uint32_t*)((char*)ptr + offset);
+            break;
+        case NNE_TYPE_STRING:
+            value.u.v_str = (char*)ptr + offset;
+            break;
+    }
+
+    return value;
+}
+
+static int nne_value_compare(struct nne_value v1, struct nne_value v2)
+{
+    if (v1.type != v2.type)
+        return -2;
+
+    switch(v1.type)
+    {
+        case NNE_TYPE_NULL:
+            return 0; // NULL equals NULL
+        case NNE_TYPE_INT8:
+            return (v1.u.v_int8 == v2.u.v_int8) ? 0 : ((v1.u.v_int8 < v2.u.v_int8) ? -1 : 1);
+        case NNE_TYPE_UINT8:
+            return (v1.u.v_uint8 == v2.u.v_uint8) ? 0 : ((v1.u.v_uint8 < v2.u.v_uint8) ? -1 : 1);
+        case NNE_TYPE_INT16:
+            return (v1.u.v_int16 == v2.u.v_int16) ? 0 : ((v1.u.v_int16 < v2.u.v_int16) ? -1 : 1);
+        case NNE_TYPE_UINT16:
+            return (v1.u.v_uint16 == v2.u.v_uint16) ? 0 : ((v1.u.v_uint16 < v2.u.v_uint16) ? -1 : 1);
+        case NNE_TYPE_INT32:
+            return (v1.u.v_int32 == v2.u.v_int32) ? 0 : ((v1.u.v_int32 < v2.u.v_int32) ? -1 : 1);
+        case NNE_TYPE_UINT32:
+            return (v1.u.v_uint32 == v2.u.v_uint32) ? 0 : ((v1.u.v_uint32 < v2.u.v_uint32) ? -1 : 1);
+        case NNE_TYPE_STRING:
+            return strcmp(v1.u.v_str, v2.u.v_str);
+    }
+
+    return -2;
+}
+
+static int md_nne_add_json_key_value_int(struct json_object *obj,
+        const char *key, int64_t value)
+{
+    struct json_object *value_obj = NULL;
+
+    value_obj = json_object_new_int64(value);
+    if (!value_obj)
+        return RETVAL_FAILURE;
+
+    json_object_object_add(obj, key, value_obj);
+    return RETVAL_SUCCESS;
+}
+
+static int md_nne_add_json_key_value_string(struct json_object *obj,
+        const char *key, const char *value)
+{
+    struct json_object *value_obj = NULL;
+
+    value_obj = json_object_new_string(value);
+    if (!value_obj)
+        return RETVAL_FAILURE;
+
+    json_object_object_add(obj, key, value_obj);
+    return RETVAL_SUCCESS;
+}
+
+static int md_nne_add_json_key_value_null(struct json_object *obj,
+        const char *key)
+{
+    json_object_object_add(obj, key, NULL);
+    return RETVAL_SUCCESS;
+}
```

Comment:
Is this needed, or can you use the fact in whaterver code that checks the json that if something is not present is it set to a default value? Or do you need to set values to NULL explicitly?